ragstack/colbertbase/colbertbase/astra_db.py

```diff
@@ -0,0 +1,179 @@
+
+        # self.cluster = Cluster(**kwargs)
+        self.cluster = Cluster(
+            cloud={
+                'secure_connect_bundle': secure_connect_bundle
+            },
+            auth_provider=PlainTextAuthProvider(
+                'token',
+                astra_token
+            )
+        )
+        self.keyspace = keyspace
+        self.session = self.cluster.connect()
+        self.session.default_timeout = timeout
+        self.verbose = verbose
+
+        print(f"set up keyspace {keyspace}, tables and indexes...")
+
+        if keyspace not in self.cluster.metadata.keyspaces.keys():
+            raise ValueError(f"Keyspace '{keyspace}' does not exist. please create it first.")
+
+        self.create_tables()
+
+        # prepare statements
+
+        chunk_counts_cql = f"""
+        SELECT COUNT(*) FROM {keyspace}.chunks
+        """
+        self.chunk_counts_stmt = self.session.prepare(chunk_counts_cql)
+
+        insert_chunk_cql = f"""
+        INSERT INTO {keyspace}.chunks (title, part, body)
+        VALUES (?, ?, ?)
+        """
+        self.insert_chunk_stmt = self.session.prepare(insert_chunk_cql)
+
+        insert_colbert_cql = f"""
+        INSERT INTO {keyspace}.colbert_embeddings (title, part, embedding_id, bert_embedding)
+        VALUES (?, ?, ?, ?)
+        """
+        self.insert_colbert_stmt = self.session.prepare(insert_colbert_cql)
+
+        query_colbert_ann_cql = f"""
+        SELECT title, part
+        FROM {keyspace}.colbert_embeddings
+        ORDER BY bert_embedding ANN OF ?
+        LIMIT ?
+        """
+        self.query_colbert_ann_stmt = self.session.prepare(query_colbert_ann_cql)
+
+        query_colbert_parts_cql = f"""
+        SELECT title, part, bert_embedding
+        FROM {keyspace}.colbert_embeddings
+        WHERE title = ? AND part = ?
+        """
+        self.query_colbert_parts_stmt = self.session.prepare(query_colbert_parts_cql)
+
+        query_part_by_pk = f"""
+        SELECT body
+        FROM {keyspace}.chunks
+        WHERE title = ? AND part = ?
+        """
+        self.query_part_by_pk_stmt = self.session.prepare(query_part_by_pk)
+
+        print("statements are prepared")
+    
+
+    def create_tables(self):
+        self.session.execute(f"""
+            use {self.keyspace};
+        """)
+        print(f"Using keyspace {self.keyspace}")
+
+        self.session.execute("""
+            CREATE TABLE IF NOT EXISTS chunks(
+                title text,
+                part int,
+                body text,
+                PRIMARY KEY (title, part)
+            );
+        """)
+        print("Created chunks table")
+
+        self.session.execute("""
+            CREATE TABLE IF NOT EXISTS colbert_embeddings (
+                title text,
+                part int,
+                embedding_id int,
+                bert_embedding vector<float, 128>,
+                PRIMARY KEY (title, part, embedding_id)
+            );
+        """)
```

Comment:
it is possible to collapse these into one table
see https://github.com/datastax-labs/colbert-wikipedia-data/blob/main/schema.cql

Reply:
@michaelsembwever in your `articles` table, is the `bert_embedding` per token's embedding? If so, would body per token or per chunk?

Reply:
> is the bert_embedding per token's embedding? 

yes

> If so, would body per token or per chunk?

body per chunk.
there's still the separate query to fetch the body.  the body is found in the rows where `bert_embedding_no=-1`

note, i probably could have saved a bit by using smallint for bert_embedding_no instead of int.  same goes for chunk_no.  we don't expect more than 32k for each.

Reply:
So the body is duplicated for each token. That's a lot of duplicated data since even with the default chunk size. There could be 160 to 220 tokens per chunk.

Reply:
No, there's no duplication. 

Example data: 

```
EXPAND set to ON
cqlsh:wikidata> select * from articles limit 3;

----------------------------+
@ Row 1
----------------------------+
wiki                       | simplewiki
 language                   | en
 title                      | Luciano Spinosi
 chunk_no                   | -1
 bert_embedding_no          | -1
 all_minilm_l6_v2_embedding | null
 bert_embedding             | null
 body                       | Luciano Spinosi\n\nLuciano Spinosi (born 9 May, 1950) is a former Italian football player. He has played for Italy national team. Strack-Zimmermann, Benjamin. "Luciano Spinosi". www.national-football-teams.com. "Luciano Spinosi - International Appearances". www.rsssf.com. This short article about a\xa0sports person can be made longer. You can help Wikipedia by adding to it.
 id                         | 182678
 revision                   | 8979208

----------------------------+
@ Row 2
----------------------------+
wiki                       | simplewiki
 language                   | en
 title                      | Luciano Spinosi
 chunk_no                   | 0
 bert_embedding_no          | -1
 all_minilm_l6_v2_embedding | [<vanilla_embedding_for_chunk_0]
 bert_embedding             | null
 body                       | Luciano Spinosi\n\nLuciano Spinosi (born 9 May, 1950) is a former Italian football player. He has played for Italy national team. Strack-Zimmermann, Benjamin. "Luciano Spinosi". www.national-football-teams.com. "Luciano Spinosi - International Appearances". www.rsssf.com. This short article about a\xa0sports person can be made longer. You can help Wikipedia by adding to it.
 id                         | 182678
 revision                   | 8979208

----------------------------+
@ Row 3
----------------------------+
 wiki                       | simplewiki
 language                   | en
 title                      | Luciano Spinosi
 chunk_no                   | 0
 bert_embedding_no          | 0
 all_minilm_l6_v2_embedding | null
 bert_embedding             | [<colbert_embedding_for_chunk_0_token_0>]
 body                       | null
 id                         | 182678
 revision                   | 8979208
```

Reply:
If a single table is requirement, there might be a way to get the actual text for the token from the ColBERT library. I have seen some functions in library that might be able to return the token. 

I am just not sure just the same table to store two different information makes sense. Is this a good practice? 
May I suggest I take this as a separate activity?

Reply:
It is good practice yes. 

The logical schema rational doesn't apply here. 

C*/DSE/Astra tables are designed by their read patterns.  And tables are intended to be sparse and have multi-dimensional growth, as opposed to SQL's one-dimensional growth.

So in the CQL paradigm this is one table that is storing only one type of information. 

There's performance wins here.  When reading the `bert_embedding_no=x` row the subsequent read of the `bert_embedding_no=-1` is likely going to be from pagecache (bc same partition key and rows close together).

Reply:
switched to a single table